IPython/html/static/widgets/js/widget.js

```diff
@@ -162,30 +188,39 @@ define(["widgets/js/manager",
             }
         },
 
+        deserialize: function(serializer, value) {
+            // given a serializer dict and a value,
+            // return a promise for the deserialized value
+            var that = this;
+            return serializer.then(function(s) {
+                if (s.deserialize) {
+                    return s.deserialize(value, that);
+                } else {
+                    return value;
+                }
+            });
+        },
+
         set_state: function (state) {
             var that = this;
             // Handle when a widget is updated via the python side.
-            return this._unpack_models(state).then(function(state) {
+            return new Promise(function(resolve, reject) {
                 that.state_lock = state;
                 try {
                     WidgetModel.__super__.set.call(that, state);
                 } finally {
                     that.state_lock = null;
                 }
+                resolve();
             }).catch(utils.reject("Couldn't set model state", true));
         },
 
         get_state: function() {
             // Get the serializable state of the model.
-            var state = this.toJSON();
-            for (var key in state) {
-                if (state.hasOwnProperty(key)) {
-                    state[key] = this._pack_models(state[key]);
-                }
-            }
-            return state;
+            // Equivalent to Backbone.Model.toJSON()
+            return _.clone(this.attributes);
```

Comment:
Doesn't this mean `get_state` will have widget model instances in it?  If it does, won't this break persistence?